staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go

```diff
@@ -129,9 +130,31 @@ func ValidateNoNewFinalizers(newFinalizers []string, oldFinalizers []string, fld
 
 // ValidateImmutableField validates the new value and the old value are deeply equal.
 func ValidateImmutableField(newVal, oldVal interface{}, fldPath *field.Path) field.ErrorList {
+	return ValidateImmutableFieldWithErr(newVal, oldVal, field.Invalid(fldPath, newVal, FieldImmutableErrorMsg))
```

Comment:
This constructs an error and calls `fldPath.String()` (which is the most expensive thing we do with a field path) on every invocation of `ValidateImmutableField`, even if the field hasn't changed and we end up discarding the error...

Accepting custom field.Error types in ValidateImmutableFieldWithErr also means we lose the ability to make the errors that come back from immutability checks consistent in a way callers could detect.

I think we should propagate `fldPath` and `FieldImmutableErrorMsg` to ValidateImmutableFieldWithErr, and construct a consistent `Invalid` error there using the passed-in message prefix, and only when the error is actually needed.

Reply:
Well actually this is what I did at the very beginning, however we found that in the original code, sometimes when the `cmp.Diff` is manually called, the error generated there is not `Invalid` but something else. Therefore if we made the ValidateImmutableFieldWithErr only generate consistent `Invalid` errors and used it to replace the `cmp.Diff`, the types of these errors would be changed, and it could cause a compatibility problem. Besides many unit tests would also be changed accordingly if we did so. So @apelisse suggested me to use the current solution here. https://github.com/kubernetes/kubernetes/pull/113963#discussion_r1028370945

Should I change it back?

Reply:
> however we found that in the original code, sometimes when the cmp.Diff is manually called, the error generated there is not Invalid but something else.

What Jordan is suggesting is to provide an API where the caller can supply all the necessary details for a custom error (`ErrorType`, prefix for the detail string) as separate parameters instead of creating an error upfront. Then the allocation only happens when validation really fails.

That makes sense to me.

Reply:
> What Jordan is suggesting is to provide an API where the caller can supply all the necessary details for a custom error (`ErrorType`, prefix for the detail string) as separate parameters instead of creating an error upfront. Then the allocation only happens when validation really fails.
> 
> That makes sense to me.

Indeed, it makes sense
The new codes have switched to the new signature